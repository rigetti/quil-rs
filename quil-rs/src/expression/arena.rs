//! First attempt to implement Expressions via an arena.
use paste::paste;
use std::{
    collections::VecDeque,
    ops::{Add, AddAssign, BitXor, BitXorAssign, Div, DivAssign, Mul, MulAssign, Sub, SubAssign},
};

// Invariants/good behavior:
// - only ever constructed or used in an arena
// - `.parent` is only ever None if this is the top node
// - `.first_child` only ever points to an index greater than where this Node is found in the arena
// - `.second_child` only ever points to an index greater than `.first_child`, and doesn't exist if
// `.first_child` doesn't.
#[derive(Clone, Debug)]
struct Node {
    kind: Kind,
    parent: Option<usize>,
    first_child: Option<usize>,
    second_child: Option<usize>,
}

#[derive(Clone, Debug)]
enum Kind {
    // Function calls
    Cis,
    Cosine,
    Exponent,
    Sine,
    SquareRoot,
    Neg,
    // Infix
    Caret,
    Plus,
    Minus,
    Slash,
    Star,
    Address(MemoryReference),
    Number(num_complex::Complex64),
    Variable(String),
}

#[derive(Clone, Debug)]
pub struct MemoryReference {
    pub name: String,
    pub index: u64,
}

// Nonempty by construction
#[derive(Clone, Debug, Default)]
struct Arena {
    nodes: VecDeque<Node>,
}

macro_rules! arena_func_call {
    ($($kind:ident),+) => {
        $(
            paste! {
                fn [<$kind:lower>](&self) -> Self {
                    let mut new = self.clone();
                    new.[<$kind:lower _mut>]();
                    new
                }
                fn [<$kind:lower _mut>](&mut self) {
                    self.shift_by_mut(1);
                    self.nodes.push_front(Node{
                        kind: Kind::$kind,
                        parent: None,
                        first_child: Some(1),
                        second_child: None,

                    })
                }
            }
        )+
    }
}

macro_rules! arena_infix {
    ($($kind:ident),+) => {
        $(
            paste! {
                fn [<$kind:lower>](&self, other: Self) -> Self {
                    let mut new = self.clone();
                    new.[<$kind:lower _mut>](other);
                    new
                }
                fn [<$kind:lower _mut>](&mut self, other: Self) {
                    let n = self.len();
                    self.shift_by_mut(1);
                    self.nodes.push_front(Node{
                        kind: Kind::$kind,
                        parent: None,
                        first_child: Some(1),
                        second_child: Some(n + 1),
                    });
                    self.extend(other.shift_by(n));
                }
            }
        )+
    }
}

macro_rules! arena_singleton {
    ($(($kind:ident, $value_type:ty)),+) => {
        $(
            paste! {
                fn [<$kind:lower>](value: $value_type) -> Self {
                    Self::new(Node {
                        kind: Kind::$kind(value),
                        parent: None,
                        first_child: None,
                        second_child: None
                    })
                }
            }
        )+
    };
}

impl Arena {
    fn extend(&mut self, other: Self) {
        self.nodes.extend(other.nodes);
    }
    fn len(&self) -> usize {
        self.nodes.len()
    }
    fn new(node: Node) -> Self {
        let mut nodes = VecDeque::with_capacity(1);
        nodes.push_back(node);
        Self { nodes }
    }
    fn shift_by_mut(&mut self, step: usize) {
        self.nodes.iter_mut().for_each(|node| {
            // set parent of base node
            if node.parent.is_none() {
                node.parent = Some(0)
            }
            node.first_child.iter_mut().for_each(|i| *i += step);
            node.second_child.iter_mut().for_each(|i| *i += step);
        });
    }
    fn shift_by(&self, step: usize) -> Self {
        let mut new = self.clone();
        new.shift_by_mut(step);
        new
    }
    arena_func_call!(Cis, Cosine, Exponent, Sine, SquareRoot, Neg);
    arena_infix!(Caret, Plus, Minus, Slash, Star);
    arena_singleton!(
        (Address, MemoryReference),
        (Number, num_complex::Complex64),
        (Variable, String)
    );
}

#[derive(Clone, Debug, Default)]
pub struct Expression {
    arena: Arena,
}

macro_rules! expr_func_call {
    ($($kind:ident),+) => {
        $(
            paste! {
                pub fn [<$kind:lower>](&self) -> Self {
                    let mut new = self.clone();
                    new.[<$kind:lower _mut>]();
                    new
                }
                pub fn [<$kind:lower _mut>](&mut self) {
                    self.arena.[<$kind:lower _mut>]();
                }
            }
        )+
    }
}

macro_rules! expr_infix {
    ($($kind:ident),+) => {
        $(
            paste! {
                pub fn [<$kind:lower>](&self, other: Self) -> Self {
                    let mut new = self.clone();
                    new.[<$kind:lower _mut>](other);
                    new
                }
                pub fn [<$kind:lower _mut>](&mut self, other: Self) {
                    self.arena.[<$kind:lower _mut>](other.arena);
                }
            }
        )+
    }
}

macro_rules! expr_singleton {
    ($(($kind:ident, $value_type:ty)),+) => {
        $(
            paste! {
                pub fn [<$kind:lower>](value: $value_type) -> Self {
                    Self { arena: Arena::[<$kind:lower>](value) }
                }
            }
        )+
    };
}

impl Expression {
    expr_func_call!(Cis, Cosine, Exponent, Sine, SquareRoot, Neg);
    expr_infix!(Caret, Plus, Minus, Slash, Star);
    expr_singleton!(
        (Address, MemoryReference),
        (Number, num_complex::Complex64),
        (Variable, String)
    );
}

macro_rules! impl_via {
    ($(($trait:ident, $method:ident)),+) => {
        $(
            paste! {
                impl $trait for Expression {
                    type Output = Self;
                    fn [<$trait:lower>](self, other: Self) -> Self::Output {
                        self.$method(other)
                    }
                }
                impl [<$trait Assign>] for Expression {
                    fn [<$trait:lower _assign>](&mut self, other: Self) {
                        self.[<$method _mut>](other)
                    }
                }

            }
        )+
    };
}

impl_via!(
    (BitXor, caret),
    (Add, plus),
    (Sub, minus),
    (Div, slash),
    (Mul, star)
);
