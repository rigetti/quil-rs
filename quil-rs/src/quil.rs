/// A trait to wrap items which represent some construct within the Quil language.
///
/// If you want to serialize an object to string and fail if it can't be represented as valid Quil, then use
/// `to_quil()`. If you want to serialize an object to string infallibly, and can tolerate invalid Quil, then
/// use `to_quil_or_debug()`.
pub trait Quil: std::fmt::Debug {
    /// Return a string in valid Quil syntax or an error if the item cannot be represented with valid Quil.
    fn to_quil(&self) -> Result<String, ToQuilError> {
        let mut buffer = String::new();
        self.write(&mut buffer, false)?;
        Ok(buffer)
    }

    /// Return a string in valid Quil syntax if possible. Any individual component of this object
    /// which cannot be represented in Quil will be replaced with a `Debug` representation of that
    /// component.
    fn to_quil_or_debug(&self) -> String {
        let mut buffer = String::new();
        let _ = self.write(&mut buffer, true);
        buffer
    }

    /// Write the Quil representation of the item to the given writer. If `fall_back_to_debug`
    /// is `true`, then it must not return an error.
    fn write(
        &self,
        writer: &mut impl std::fmt::Write,
        fall_back_to_debug: bool,
    ) -> Result<(), ToQuilError>;
}

pub type ToQuilResult<T> = Result<T, ToQuilError>;

/// Errors which can occur when converting a Quil item to a string.
#[derive(Debug, thiserror::Error, PartialEq)]
#[non_exhaustive]
pub enum ToQuilError {
    #[error("Failed to write Quil: {0}")]
    FormatError(#[from] std::fmt::Error),
    #[error("Label has not yet been resolved")]
    UnresolvedLabelPlaceholder,
    #[error("Qubit has not yet been resolved")]
    UnresolvedQubitPlaceholder,
}

/// Write an iterator of Quil items to the given writer, joined with the provided `joiner`.
pub(crate) fn write_join_quil<'i, I, T>(
    writer: &mut impl std::fmt::Write,
    fall_back_to_debug: bool,
    values: I,
    joiner: &str,
    prefix: &str,
) -> Result<(), ToQuilError>
where
    I: IntoIterator<Item = &'i T>,
    T: Quil + 'i,
{
    let mut iter = values.into_iter();
    if let Some(first) = iter.next() {
        write!(writer, "{prefix}")?;
        first.write(writer, fall_back_to_debug)?;

        for value in iter {
            write!(writer, "{joiner}{prefix}")?;
            value.write(writer, fall_back_to_debug)?;
        }
    }
    Ok(())
}
