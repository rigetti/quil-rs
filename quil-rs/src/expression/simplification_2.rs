/// Complex machinery for simplifying [`Expression`]s.
use crate::expression::{
    imag, is_small, real, Expression, ExpressionFunction, FunctionCallExpression, InfixExpression,
    InfixOperator, PrefixExpression, PrefixOperator,
};
use std::f64::consts::PI;

/// Simplify an [`Expression`].
pub(super) fn run(expression: Expression) -> Expression {
    simplify(expression)
}

fn simplify(e: Expression) -> Expression {
    match e {
        Expression::Address(_)
        | Expression::Number(_)
        | Expression::PiConstant
        | Expression::Variable(_) => e,
        Expression::FunctionCall(FunctionCallExpression {
            function,
            expression,
        }) => simplify_function_call(function, *expression),
        Expression::Infix(InfixExpression {
            left,
            operator,
            right,
        }) => simplify_infix(*left, operator, *right),
        Expression::Prefix(PrefixExpression {
            operator,
            expression,
        }) => simplify_prefix(operator, *expression),
    }
}

fn simplify_function_call(func: ExpressionFunction, expr: Expression) -> Expression {
    // Evaluate numbers and π; pass through otherwise
    match (func, simplify(expr)) {
        (ExpressionFunction::Cis, Expression::Number(x)) => {
            // num_complex::Complex::<f64>::cis only accpets f64 :-(
            Expression::Number(x.cos() + imag!(1f64) * x.sin())
        }
        (ExpressionFunction::Cis, Expression::PiConstant) => Expression::Number(real!(-1f64)),
        (ExpressionFunction::Cosine, Expression::Number(x)) => Expression::Number(x.cos()),
        (ExpressionFunction::Cosine, Expression::PiConstant) => Expression::Number(real!(-1f64)),
        (ExpressionFunction::Exponent, Expression::Number(x)) => Expression::Number(x.exp()),
        (ExpressionFunction::Exponent, Expression::PiConstant) => {
            Expression::Number(real!(PI).exp())
        }
        (ExpressionFunction::Sine, Expression::Number(x)) => Expression::Number(x.sin()),
        (ExpressionFunction::Sine, Expression::PiConstant) => Expression::Number(real!(PI).sin()),
        (ExpressionFunction::SquareRoot, Expression::Number(x)) => Expression::Number(x.sqrt()),
        (ExpressionFunction::SquareRoot, Expression::PiConstant) => {
            Expression::Number(real!(PI).sqrt())
        }
        (function, expression) => Expression::FunctionCall(FunctionCallExpression {
            function,
            expression: expression.into(),
        }),
    }
}

fn simplify_infix(l: Expression, op: InfixOperator, r: Expression) -> Expression {
    match (simplify(l), op, simplify(r)) {
        (left, operator, right) => Expression::Infix(InfixExpression {
            left: left.into(),
            operator,
            right: right.into(),
        }),
    }
}

fn simplify_prefix(op: PrefixOperator, expr: Expression) -> Expression {
    // Remove +; push - into numbers & π; pass through otherwise
    match (op, simplify(expr)) {
        (PrefixOperator::Plus, expression) => expression,
        (PrefixOperator::Minus, Expression::Number(x)) => Expression::Number(-x),
        (PrefixOperator::Minus, Expression::PiConstant) => Expression::Number(real!(-PI)),
        (operator, expression) => Expression::Prefix(PrefixExpression {
            operator,
            expression: expression.into(),
        }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::str::FromStr;

    macro_rules! test_simplify {
        ($name:ident, $input:expr, $expected:expr$(,)?) => {
            #[test]
            fn $name() {
                let parsed_input = Expression::from_str($input);
                assert!(parsed_input.is_ok(), "Parsing input `{}` failed!", $input);
                let parsed_expected = Expression::from_str($expected);
                assert!(
                    parsed_expected.is_ok(),
                    "Parsing expected expression `{}` failed!",
                    $expected
                );
                let output = simplify(parsed_input.unwrap());
                assert_eq!(
                    parsed_expected.unwrap(),
                    output,
                    "Simplifying `{}` yielded `{}` instead of the expected `{}`",
                    $input,
                    output,
                    $expected
                );
            }
        };
    }

    test_simplify! {
        docstring_example,
        "cos(2 * pi) + 2",
        "3"
    }

    test_simplify! {
        issue_208_1,
        "0 * theta[0]",
        "0"
    }

    test_simplify! {
        issue_208_2,
        "theta[0] / 1",
        "theta[0]"
    }

    test_simplify! {
        issue_208_3,
        "(theta[0] * 5) / 5",
        "theta"
    }

    test_simplify! {
        memory_ref,
        "theta[0]",
        "theta[0]"
    }

    test_simplify! {
        var,
        "%foo",
        "%foo"
    }

    test_simplify! {
        prefix_neg,
        "-(-1)",
        "1"
    }

    test_simplify! {
        neg_sub,
        "-(1 - 2)",
        "1"
    }

    test_simplify! {
        neg_imag,
        "-(9.48e42i)",
        "-9.48e42i"
    }

    test_simplify! {
        pow_neg_address,
        "(-(9.48e42i))^A[9]",
        "(-9.48e42i)^A[9]",
    }

    test_simplify! {
        fold_constant_mul,
        "2 * pi",
        "6.283185307179586"
    }

    test_simplify! {
        fold_constant_mul_div,
        "(2 * pi) / 6.283185307179586",
        "1"
    }

    test_simplify! {
        fold_constant_mul_div_with_ref,
        "((a[0] * 2) * pi) / 6.283185307179586",
        "a[0]"
    }
}
