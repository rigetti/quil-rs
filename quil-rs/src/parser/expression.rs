// Copyright 2021 Rigetti Computing
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use nom::combinator::opt;

use crate::expression::{FunctionCallExpression, InfixExpression, PrefixExpression};
use crate::parser::InternalParserResult;
use crate::{
    expected_token,
    expression::{Expression, ExpressionFunction, InfixOperator, PrefixOperator},
    imag,
    instruction::MemoryReference,
    parser::common::parse_memory_reference_with_brackets,
    token, unexpected_eof,
};

use super::lexer::{Operator, Token};
use super::ParserInput;

#[derive(Debug, PartialEq, PartialOrd)]
enum Precedence {
    Lowest,
    Sum,
    Product,
    Exponentiation,
    Call,
}

impl From<&Token> for Precedence {
    fn from(token: &Token) -> Self {
        match token {
            Token::Operator(operator) => Self::from(operator),
            // TODO: Is this used?
            Token::LParenthesis => Precedence::Call,
            _ => Precedence::Lowest,
        }
    }
}

impl From<&Operator> for Precedence {
    fn from(operator: &Operator) -> Self {
        match operator {
            Operator::Plus | Operator::Minus => Precedence::Sum,
            Operator::Star | Operator::Slash => Precedence::Product,
            Operator::Caret => Precedence::Exponentiation,
        }
    }
}

fn get_precedence(input: ParserInput) -> Precedence {
    match super::first_token(input) {
        Some(v) => Precedence::from(v),
        None => Precedence::Lowest,
    }
}

/// Parse an expression at the head of the current input, for as long as the expression continues.
/// Return an error only if the first token(s) do not form an expression.
pub(crate) fn parse_expression(input: ParserInput) -> InternalParserResult<Expression> {
    parse(input, Precedence::Lowest)
}

/// Recursively parse an expression as long as operator precedence is satisfied.
fn parse(input: ParserInput, precedence: Precedence) -> InternalParserResult<Expression> {
    let (input, prefix) = opt(parse_prefix)(input)?;
    let (mut input, mut left) = match super::split_first_token(input) {
        None => unexpected_eof!(input),
        Some((Token::Integer(value), remainder)) => {
            let (remainder, imaginary) = opt(parse_i)(remainder)?;
            match imaginary {
                None => Ok((remainder, Expression::Number(crate::real!(*value as f64)))),
                Some(_) => Ok((remainder, Expression::Number(crate::imag!(*value as f64)))),
            }
        }
        Some((Token::Float(value), remainder)) => {
            let (remainder, imaginary) = opt(parse_i)(remainder)?;
            match imaginary {
                None => Ok((remainder, Expression::Number(crate::real!(*value)))),
                Some(_) => Ok((remainder, Expression::Number(crate::imag!(*value)))),
            }
        }
        Some((Token::Variable(name), remainder)) => {
            Ok((remainder, Expression::Variable(name.clone())))
        }
        Some((Token::Identifier(_), _)) => parse_expression_identifier(input),
        Some((Token::LParenthesis, remainder)) => parse_grouped_expression(remainder),
        Some((token, _)) => {
            expected_token!(input, token, "expression".to_owned())
        }
    }?;

    if let Some(prefix) = prefix {
        left = Expression::Prefix(PrefixExpression {
            operator: prefix,
            expression: Box::new(left),
        });
    }

    while get_precedence(input) > precedence {
        match super::first_token(input) {
            None => return Ok((input, left)),
            Some(Token::Operator(_)) => {
                let (remainder, expression) = parse_infix(input, left)?;
                left = expression;
                input = remainder;
            }
            Some(_) => return Ok((input, left)),
        }
    }

    Ok((input, left))
}

/// Returns successfully if the head of input is the identifier `i`, returns error otherwise.
fn parse_i(input: ParserInput) -> InternalParserResult<()> {
    match super::split_first_token(input) {
        None => unexpected_eof!(input),
        Some((Token::Identifier(v), remainder)) if v == "i" => Ok((remainder, ())),
        Some((other_token, _)) => expected_token!(input, other_token, "i".to_owned()),
    }
}

/// Given an expression function, parse the expression within its parentheses.
fn parse_function_call<'a>(
    input: ParserInput<'a>,
    function: ExpressionFunction,
) -> InternalParserResult<'a, Expression> {
    let (input, _) = token!(LParenthesis)(input)?;
    let (input, expression) = parse(input, Precedence::Lowest)?; // TODO: different precedence?
    let (input, _) = token!(RParenthesis)(input)?;
    Ok((
        input,
        Expression::FunctionCall(FunctionCallExpression {
            function,
            expression: Box::new(expression),
        }),
    ))
}

/// Identifiers have to be handled specially because some have special meaning.
///
/// By order of precedence:
///
/// 1. Memory references with brackets
/// 2. Special function and constant identifiers
/// 3. Anything else is considered to be a memory reference without index brackets
fn parse_expression_identifier(input: ParserInput) -> InternalParserResult<Expression> {
    let (input, memory_reference) = opt(parse_memory_reference_with_brackets)(input)?;
    if let Some(memory_reference) = memory_reference {
        return Ok((input, Expression::Address(memory_reference)));
    }

    match super::split_first_token(input) {
        None => unexpected_eof!(input),
        Some((Token::Identifier(ident), remainder)) => match ident.to_lowercase().as_str() {
            "cis" => parse_function_call(remainder, ExpressionFunction::Cis),
            "cos" => parse_function_call(remainder, ExpressionFunction::Cosine),
            "exp" => parse_function_call(remainder, ExpressionFunction::Exponent),
            "i" => Ok((remainder, Expression::Number(imag!(1f64)))),
            "pi" => Ok((remainder, Expression::PiConstant)),
            "sin" => parse_function_call(remainder, ExpressionFunction::Sine),
            "sqrt" => parse_function_call(remainder, ExpressionFunction::SquareRoot),
            name => Ok((
                remainder,
                Expression::Address(MemoryReference {
                    name: name.to_owned(),
                    index: 0,
                }),
            )),
        },
        Some((other_token, _)) => expected_token!(input, other_token, "identifier".to_owned()),
    }
}

/// To be called following an opening parenthesis, this will parse the expression to its end
/// and then expect a closing right parenthesis.
fn parse_grouped_expression(input: ParserInput) -> InternalParserResult<Expression> {
    let (input, expression) = parse(input, Precedence::Lowest)?;
    match super::split_first_token(input) {
        None => unexpected_eof!(input),
        Some((Token::RParenthesis, remainder)) => Ok((remainder, expression)),
        Some((other_token, _)) => {
            expected_token!(input, other_token, "right parenthesis".to_owned())
        }
    }
}

/// Parse an infix operator and then the expression to the right of the operator, and return the
/// resulting infixed expression.
fn parse_infix(input: ParserInput, left: Expression) -> InternalParserResult<Expression> {
    match super::split_first_token(input) {
        None => unexpected_eof!(input),
        Some((Token::Operator(token_operator), remainder)) => {
            let expression_operator = match token_operator {
                Operator::Plus => InfixOperator::Plus,
                Operator::Minus => InfixOperator::Minus,
                Operator::Caret => InfixOperator::Caret,
                Operator::Slash => InfixOperator::Slash,
                Operator::Star => InfixOperator::Star,
            };
            let precedence = Precedence::from(token_operator);
            let (remainder, right) = parse(remainder, precedence)?;
            let infix_expression = Expression::Infix(InfixExpression {
                left: Box::new(left),
                operator: expression_operator,
                right: Box::new(right),
            });
            Ok((remainder, infix_expression))
        }
        Some((other_token, _)) => expected_token!(input, other_token, "infix operator".to_owned()),
    }
}

/// Return the prefix operator at the beginning of the input, if any.
fn parse_prefix(input: ParserInput) -> InternalParserResult<PrefixOperator> {
    match super::split_first_token(input) {
        None => unexpected_eof!(input),
        Some((Token::Operator(Operator::Minus), remainder)) => {
            Ok((remainder, PrefixOperator::Minus))
        }
        Some((other_token, _)) => expected_token!(input, other_token, "prefix operator".to_owned()),
    }
}

#[cfg(test)]
mod tests {
    use crate::expression::{
        Expression, ExpressionFunction, FunctionCallExpression, InfixExpression, InfixOperator,
        PrefixExpression, PrefixOperator,
    };
    use crate::instruction::MemoryReference;
    use crate::parser::lexer::lex;
    use crate::quil::Quil;
    use crate::{imag, real};

    use nom_locate::LocatedSpan;

    use super::parse_expression;

    macro_rules! test {
        ($name: ident, $parser: ident, $input: expr, $expected: expr) => {
            #[test]
            fn $name() {
                let input = LocatedSpan::new($input);
                let tokens = lex(input).unwrap();
                let (remainder, parsed) = $parser(&tokens).unwrap();
                assert_eq!(remainder.len(), 0);
                assert_eq!(parsed, $expected);
            }
        };
    }

    // given a function and vector of (input, expected) tuples, invoke the function on each and
    // panic on the first mismatch
    fn compare(cases: Vec<(&str, Expression)>) {
        for case in cases {
            let input = LocatedSpan::new(case.0);
            let tokens = lex(input).unwrap();
            let (remainder, parsed) = parse_expression(&tokens).unwrap();
            assert_eq!(remainder.len(), 0);
            assert_eq!(case.1, parsed);
        }
    }

    // Round-trip expressions to validate parsing & display
    #[test]
    fn display() {
        let cases = vec![
            "pi",
            "sin(pi)",
            "1+(2*3)",
            "(1+2)*3",
            "%theta",
            "cis(%theta)",
            "%a+%b",
            "(pi/2)+(1*theta[0])",
            "3 - -2",
        ];

        for case in cases {
            let input = LocatedSpan::new(case);
            let tokens = lex(input).unwrap();
            let (remainder, parsed) = parse_expression(&tokens).unwrap();
            assert_eq!(remainder.len(), 0);
            assert_eq!(parsed.to_quil_or_debug(), case);
        }
    }

    test!(
        function_call,
        parse_expression,
        "sin(1)",
        Expression::FunctionCall(FunctionCallExpression {
            function: ExpressionFunction::Sine,
            expression: Box::new(Expression::Number(real!(1f64))),
        })
    );

    test!(
        nested_function_call,
        parse_expression,
        "sin(sin(1))",
        Expression::FunctionCall(FunctionCallExpression {
            function: ExpressionFunction::Sine,
            expression: Box::new(Expression::FunctionCall(FunctionCallExpression {
                function: ExpressionFunction::Sine,
                expression: Box::new(Expression::Number(real!(1f64))),
            })),
        })
    );

    test!(
        simple_infix,
        parse_expression,
        "1+2",
        Expression::Infix(InfixExpression {
            left: Box::new(Expression::Number(real!(1f64))),
            operator: InfixOperator::Plus,
            right: Box::new(Expression::Number(real!(2f64))),
        })
    );

    test!(
        infix_with_function_call,
        parse_expression,
        "-i*sin(%theta/2)",
        Expression::Infix(InfixExpression {
            left: Box::new(Expression::Prefix(PrefixExpression {
                operator: PrefixOperator::Minus,
                expression: Box::new(Expression::Number(imag!(1f64))),
            })),
            operator: InfixOperator::Star,
            right: Box::new(Expression::FunctionCall(FunctionCallExpression {
                function: ExpressionFunction::Sine,
                expression: Box::new(Expression::Infix(InfixExpression {
                    left: Box::new(Expression::Variable("theta".to_owned())),
                    operator: InfixOperator::Slash,
                    right: Box::new(Expression::Number(real!(2f64))),
                })),
            })),
        })
    );

    test!(
        infix_parenthesized,
        parse_expression,
        "(1+2i)*%a",
        Expression::Infix(InfixExpression {
            left: Box::new(Expression::Infix(InfixExpression {
                left: Box::new(Expression::Number(real!(1f64))),
                operator: InfixOperator::Plus,
                right: Box::new(Expression::Number(imag!(2f64))),
            })),
            operator: InfixOperator::Star,
            right: Box::new(Expression::Variable("a".to_owned())),
        })
    );

    test!(
        infix_with_infix_operands_implicit_precedence,
        parse_expression,
        "pi/2 + 2*theta[0]",
        Expression::Infix(InfixExpression {
            left: Box::new(Expression::Infix(InfixExpression {
                left: Box::new(Expression::PiConstant),
                operator: InfixOperator::Slash,
                right: Box::new(Expression::Number(real!(2f64))),
            })),
            operator: InfixOperator::Plus,
            right: Box::new(Expression::Infix(InfixExpression {
                left: Box::new(Expression::Number(real!(2f64))),
                operator: InfixOperator::Star,
                right: Box::new(Expression::Address(MemoryReference {
                    name: "theta".to_string(),
                    index: 0,
                })),
            })),
        })
    );

    test!(
        infix_minus_negative,
        parse_expression,
        "-3 - -2",
        Expression::Infix(InfixExpression {
            left: Box::new(Expression::Prefix(PrefixExpression {
                operator: PrefixOperator::Minus,
                expression: Box::new(Expression::Number(real!(3f64))),
            })),
            operator: InfixOperator::Minus,
            right: Box::new(Expression::Prefix(PrefixExpression {
                operator: PrefixOperator::Minus,
                expression: Box::new(Expression::Number(real!(2f64)))
            }))
        })
    );

    #[test]
    fn parenthetical() {
        let cases = vec![
            (
                "1 + ( 2 + 3 )",
                Expression::Infix(InfixExpression {
                    left: Box::new(Expression::Number(real!(1f64))),
                    operator: InfixOperator::Plus,
                    right: Box::new(Expression::Infix(InfixExpression {
                        left: Box::new(Expression::Number(real!(2f64))),
                        operator: InfixOperator::Plus,
                        right: Box::new(Expression::Number(real!(3f64))),
                    })),
                }),
            ),
            (
                "1+(2+3)",
                Expression::Infix(InfixExpression {
                    left: Box::new(Expression::Number(real!(1f64))),
                    operator: InfixOperator::Plus,
                    right: Box::new(Expression::Infix(InfixExpression {
                        left: Box::new(Expression::Number(real!(2f64))),
                        operator: InfixOperator::Plus,
                        right: Box::new(Expression::Number(real!(3f64))),
                    })),
                }),
            ),
            (
                "(1+2)+3",
                Expression::Infix(InfixExpression {
                    left: Box::new(Expression::Infix(InfixExpression {
                        left: Box::new(Expression::Number(real!(1f64))),
                        operator: InfixOperator::Plus,
                        right: Box::new(Expression::Number(real!(2f64))),
                    })),
                    operator: InfixOperator::Plus,
                    right: Box::new(Expression::Number(real!(3f64))),
                }),
            ),
            (
                "(((cos(((pi))))))",
                Expression::FunctionCall(FunctionCallExpression {
                    function: ExpressionFunction::Cosine,
                    expression: Box::new(Expression::PiConstant),
                }),
            ),
        ];

        compare(cases);
    }

    #[test]
    fn pi() {
        let cases = vec![("pi", Expression::PiConstant)];

        compare(cases);
    }

    #[test]
    fn variable() {
        let cases = vec![
            ("%theta", Expression::Variable("theta".to_owned())),
            ("%pi", Expression::Variable("pi".to_owned())),
            ("%sin", Expression::Variable("sin".to_owned())),
        ];

        compare(cases);
    }
}
