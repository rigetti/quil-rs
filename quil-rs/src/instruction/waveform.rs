use std::collections::HashMap;

use crate::{
    expression::Expression,
    quil::{write_join_quil, Quil},
};

use super::write_parameter_string;

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Waveform {
    pub matrix: Vec<Expression>,
    pub parameters: Vec<String>,
}

impl Waveform {
    pub fn new(matrix: Vec<Expression>, parameters: Vec<String>) -> Self {
        Self { matrix, parameters }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct WaveformDefinition {
    pub name: String,
    pub definition: Waveform,
}

impl WaveformDefinition {
    pub fn new(name: String, definition: Waveform) -> Self {
        Self { name, definition }
    }
}

impl Quil for WaveformDefinition {
    fn write(
        &self,
        f: &mut impl std::fmt::Write,
        fall_back_to_debug: bool,
    ) -> crate::quil::ToQuilResult<()> {
        write!(f, "DEFWAVEFORM {}", self.name)?;
        write_parameter_string(f, &self.definition.parameters)?;
        write!(f, ":\n\t")?;
        write_join_quil(f, fall_back_to_debug, &self.definition.matrix, ", ", "")
            .map_err(Into::into)
    }
}

#[cfg(test)]
mod test_waveform_definition {
    use super::{Waveform, WaveformDefinition};
    use crate::expression::Expression;
    use crate::quil::Quil;
    use crate::real;

    use insta::assert_snapshot;
    use rstest::rstest;

    #[rstest]
    #[case(
        "Simple WaveformDefinition",
        WaveformDefinition {
            name: "WF".to_string(),
            definition: Waveform {
                matrix: vec![
                    Expression::Number(real!(0.0)),
                    Expression::Number(real!(0.0)),
                    Expression::Number(real!(0.00027685415721916584))
                ],
                parameters: vec!["theta".to_string()]
            }
        }
    )]
    fn test_display(#[case] description: &str, #[case] waveform_def: WaveformDefinition) {
        insta::with_settings!({
            snapshot_suffix => description,
        }, {
            assert_snapshot!(waveform_def.to_quil_or_debug())
        })
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WaveformInvocation {
    pub name: String,
    pub parameters: HashMap<String, Expression>,
}

impl WaveformInvocation {
    pub fn new(name: String, parameters: HashMap<String, Expression>) -> Self {
        Self { name, parameters }
    }
}

impl Quil for WaveformInvocation {
    fn write(
        &self,
        f: &mut impl std::fmt::Write,
        _fall_back_to_debug: bool,
    ) -> crate::quil::ToQuilResult<()> {
        let mut key_value_pairs = self
            .parameters
            .iter()
            .collect::<Vec<(&String, &Expression)>>();

        key_value_pairs.sort_by(|(k1, _), (k2, _)| k1.cmp(k2));

        if key_value_pairs.is_empty() {
            write!(f, "{}", self.name,)?;
        } else {
            write!(
                f,
                "{}({})",
                self.name,
                key_value_pairs
                    .iter()
                    .map(|(k, v)| format!("{k}: {}", v.to_quil_or_debug()))
                    .collect::<Vec<String>>()
                    .join(", ")
            )?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod waveform_invocation_tests {
    use std::collections::HashMap;

    use crate::{instruction::WaveformInvocation, quil::Quil};

    #[test]
    fn format_no_parameters() {
        let invocation = WaveformInvocation {
            name: "CZ".into(),
            parameters: HashMap::new(),
        };
        assert_eq!(invocation.to_quil_or_debug(), "CZ".to_string());
    }
}
