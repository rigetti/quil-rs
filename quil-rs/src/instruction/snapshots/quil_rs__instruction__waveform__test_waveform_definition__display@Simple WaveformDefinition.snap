---
source: quil-rs/src/instruction/waveform.rs
expression: waveform_def.to_string()
---
DEFWAVEFORM WF(%theta):
	0, 0, 0.00027685415721916584
